source/core/Variants.cpp

```diff
@@ -0,0 +1,340 @@
+
+SDG
+*/
+
+/*! \file
+*/
+
+#include "TypeDefiner.h"
+#include "ExecutionContext.h"
+#include "TypeAndDataManager.h"
+#include "TDCodecVia.h"
+#include <limits>
+#include <map>
+
+namespace Vireo
+{
+class VariantAttributeManager {
+ public:
+    using AttributeMapType = std::map<StringRef, StaticTypeAndDataPointer, StringRefCmp>;
+    using VariantToAttributeMapType = std::map< DataPointer, AttributeMapType *>;
+    VariantAttributeManager(const VariantAttributeManager&) = delete;
+    VariantAttributeManager& operator=(const VariantAttributeManager&) = delete;
+
+ private:
+    VariantToAttributeMapType _variantToAttributeMap;
+    VariantAttributeManager() = default;
+
+ public:
+    static VariantAttributeManager& Instance() {
+        static VariantAttributeManager _instance;
+        return _instance;
+    }
+
+    const VariantToAttributeMapType& GetVariantToAttributeMap() const {
+        return _variantToAttributeMap;
+    }
+
+    VariantToAttributeMapType& GetVariantToAttributeMap() {
+        return _variantToAttributeMap;
+    }
+
+    ~VariantAttributeManager() {
+        for (auto entry : _variantToAttributeMap) {
+            entry.second->clear();
+        }
+        _variantToAttributeMap.clear();
+    }
+};
+
+//------------------------------------------------------------
+struct DataToVariantParamBlock : public VarArgInstruction
+{
+    _ParamImmediateDef(StaticTypeAndData, InputData);
+    _ParamDef(TypeRef, OutputVariant);
+    NEXT_INSTRUCTION_METHODV()
+};
+
+// Convert data of any type to variant
+VIREO_FUNCTION_SIGNATUREV(DataToVariant, DataToVariantParamBlock)
+{
+    TypeRef inputType = _ParamImmediate(InputData._paramType);
+    TypeManagerRef tm = THREAD_TADM();
+
+    TypeRef variantType = DefaultValueType::New(tm, inputType, true);
+    variantType->CopyData(_ParamImmediate(InputData._pData), variantType->Begin(kPAWrite));
+    _Param(OutputVariant) = variantType;
+    return _NextInstruction();
+}
+
+//------------------------------------------------------------
+struct VariantToDataParamBlock : public VarArgInstruction
+{
+    _ParamDef(TypeRef, VariantType);
+    _ParamDef(ErrorCluster, ErrorClust);
+    _ParamImmediateDef(StaticTypeAndData, DestData);
+    NEXT_INSTRUCTION_METHODV()
+};
+
+// Convert variant to data of given type. Error if the data types don't match
+VIREO_FUNCTION_SIGNATUREV(VariantToData, VariantToDataParamBlock)
+{
+    ErrorCluster *errPtr = _ParamPointer(ErrorClust);
+    if (!errPtr || !errPtr->status) {
+        TypeRef variantType = _Param(VariantType);
+        TypeRef destType = _ParamImmediate(DestData._paramType);
+
+        if (variantType->IsA(destType)) {
+            void* destData = _ParamImmediate(DestData)._pData;
+            variantType->CopyData(variantType->Begin(kPARead), destData);
+        } else {
+            if (errPtr) {
+                errPtr->SetError(true, 91, "Variant To Data");
+            }
+        }
+    }
```

Comment:
The primitive write default data to its output when an error is passed in, or when the destType doesn't match the variant's contained type.